test/test_cubic_b_spline.cpp

```diff
@@ -0,0 +1,217 @@
+#define BOOST_TEST_MODULE test_cubic_b_spline
+
+#include <random>
+#include <boost/random/uniform_real_distribution.hpp>
+#include <boost/type_index.hpp>
+#include <boost/test/included/unit_test.hpp>
+#include <boost/test/floating_point_comparison.hpp>
+#include <boost/math/interpolators/cubic_b_spline.hpp>
+
+
+
+using boost::math::constants::third;
+using boost::math::constants::half;
+
+template<class Real>
+void test_b3_spline()
+{
+    std::cout << "Testing evaluation of spline basis functions on type " << boost::typeindex::type_id<Real>().pretty_name() << "\n";
+    // Outside the support:
+    auto eps = std::numeric_limits<Real>::epsilon();
+    BOOST_CHECK_SMALL(boost::math::b3_spline<Real>(2.5), (Real) 0);
+    BOOST_CHECK_SMALL(boost::math::b3_spline<Real>(-2.5), (Real) 0);
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(2.5), (Real) 0);
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(-2.5), (Real) 0);
+
+    // On the boundary of support:
+    BOOST_CHECK_SMALL(boost::math::b3_spline<Real>(2), (Real) 0);
+    BOOST_CHECK_SMALL(boost::math::b3_spline<Real>(-2), (Real) 0);
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(2), (Real) 0);
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(-2), (Real) 0);
+
+    // Special values:
+    BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>(-1), third<Real>()*half<Real>(), eps);
+    BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>( 1), third<Real>()*half<Real>(), eps);
+    BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>(0), 2*third<Real>(), eps);
+
+    BOOST_CHECK_CLOSE(boost::math::b3_spline_prime<Real>(-1), half<Real>(), eps);
+    BOOST_CHECK_CLOSE(boost::math::b3_spline_prime<Real>( 1), -half<Real>(), eps);
+    BOOST_CHECK_SMALL(boost::math::b3_spline_prime<Real>(0), eps);
+
+    // Properties: B3 is an even function, B3' is an odd function.
+    for (size_t i = 1; i < 200; ++i)
+    {
+        auto arg = i*0.01;
+        BOOST_CHECK_CLOSE(boost::math::b3_spline<Real>(arg), boost::math::b3_spline<Real>(arg), eps);
+        BOOST_CHECK_CLOSE(boost::math::b3_spline_prime<Real>(-arg), -boost::math::b3_spline_prime<Real>(arg), eps);
+    }
+
+}
+/*
+ * This test ensures that the interpolant s(x_j) = f(x_j) at all grid points.
+ */
+template<class Real>
+void test_interpolation_condition()
+{
+    std::cout << "Testing interpolation condition for cubic b splines on type " << boost::typeindex::type_id<Real>().pretty_name()  << "\n";
+    std::random_device rd;
```

Comment:
I removed most of the randomness from this test, however I think that testing the interpolation condition with random points is a very strong one. I've reduced the precision so that I couldn't get any failures on many runs.